From b4589d42bcb30445df52028e5514475b381b4157 Mon Sep 17 00:00:00 2001
From: Alain Volmat <alain.volmat@foss.st.com>
Date: Sat, 23 Mar 2024 20:29:47 +0100
Subject: [PATCH 1032/1141] media: dcmipp: add missing
 v4l2_pipeline_link_notify handler

Add v4l2_pipeline_link_notify handler in order to properly
handle the entity use_count counter.

Change-Id: I4f483e7bf02070ebe0ceb0661a74c875da440ec6
Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/369265
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Philippe CORNU <philippe.cornu@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Philippe CORNU <philippe.cornu@foss.st.com>
---
 .../media/platform/st/stm32/stm32-dcmipp/dcmipp-bytecap.c   | 1 -
 drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-core.c  | 6 ++++++
 .../media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelcap.c  | 1 -
 .../media/platform/st/stm32/stm32-dcmipp/dcmipp-statcap.c   | 1 -
 4 files changed, 6 insertions(+), 3 deletions(-)

diff --git a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-bytecap.c b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-bytecap.c
index f5c49aa6c61b..a4a8782d294a 100644
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-bytecap.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-bytecap.c
@@ -331,7 +331,6 @@ static int dcmipp_bytecap_enum_framesizes(struct file *file, void *fh,
 	return 0;
 }
 
-/* TODO - based on the explanation text, should also use v4l2_pipeline_link_notify */
 static int dcmipp_bytecap_open(struct file *file)
 {
 	struct dcmipp_bytecap_device *vcap = video_drvdata(file);
diff --git a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-core.c b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-core.c
index 55df89ba97a0..f89ee41a78f6 100644
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-core.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-core.c
@@ -21,6 +21,7 @@
 #include <media/media-device.h>
 #include <media/v4l2-device.h>
 #include <media/v4l2-fwnode.h>
+#include <media/v4l2-mc.h>
 
 #include "dcmipp-common.h"
 
@@ -463,6 +464,10 @@ static int dcmipp_graph_init(struct dcmipp_device *dcmipp)
 	return 0;
 }
 
+static const struct media_device_ops dcmipp_media_ops = {
+	.link_notify = v4l2_pipeline_link_notify,
+};
+
 static int dcmipp_probe(struct platform_device *pdev)
 {
 	struct dcmipp_device *dcmipp;
@@ -571,6 +576,7 @@ static int dcmipp_probe(struct platform_device *pdev)
 	snprintf(dcmipp->mdev.bus_info, sizeof(dcmipp->mdev.bus_info),
 		 "platform:%s", DCMIPP_PDEV_NAME);
 	dcmipp->mdev.dev = &pdev->dev;
+	dcmipp->mdev.ops = &dcmipp_media_ops;
 	media_device_init(&dcmipp->mdev);
 
 	pm_runtime_enable(dcmipp->dev);
diff --git a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelcap.c b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelcap.c
index a2f5fa0fe8c3..2e734c1e0170 100644
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelcap.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelcap.c
@@ -436,7 +436,6 @@ static int dcmipp_pixelcap_enum_framesizes(struct file *file, void *fh,
 	return 0;
 }
 
-/* TODO - based on the explanation text, should also use v4l2_pipeline_link_notify */
 static int dcmipp_pixelcap_open(struct file *file)
 {
 	struct dcmipp_pixelcap_device *vcap = video_drvdata(file);
diff --git a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-statcap.c b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-statcap.c
index ddd5896bfaeb..81b6fbf6de04 100644
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-statcap.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-statcap.c
@@ -161,7 +161,6 @@ static int dcmipp_statcap_enum_fmt_meta_cap(struct file *file, void *priv,
 	return 0;
 }
 
-/* TODO - based on the explanation text, should also use v4l2_pipeline_link_notify */
 static int dcmipp_statcap_open(struct file *file)
 {
 	struct dcmipp_statcap_device *vcap = video_drvdata(file);
-- 
2.39.2

